reps/2022-09-19-ray-on-spark.md

```diff
@@ -0,0 +1,286 @@
+## Summary
+### General Motivation
+
+Supporting running Ray applications on spark cluster / databricks runtime,
+Spark is a popular distributed computing framework and is used widely,
+If supporting running Ray applications on spark cluster, user don't need to
+setup a standalone ray cluster and it allows ray workloads and spark workloads
+runs together.
+
+#### Key requirements:
+- Ray application resources (cpu/gpu/memory) allocation must respect spark job resource allocation.
+- Little overhead compared with native Ray cluster.
+- (Optional) Ray UI portal support.
+
+### Should this change be within `ray` or outside?
+
+Yes. For better code maintemance.
+
+## Stewardship
+### Required Reviewers
+
+- @jjyao
+- @ericl
+
+### Shepherd of the Proposal (should be a senior committer)
+
+- @jjyao
+
+## Design and Architecture
+
+### Prerequisites
+
+- The user have an active Spark cluster (apache/spark >= 3.3)
+
+- The user must install Ray packages on every node of the Spark cluster, e.g. via:
+
+  ```
+  pip install ray
+  ```
+
+  If the user's workload requires Ray add-ons, they must also be installed, e.g. via
+
+  ```
+  pip install "ray[debug,dashboard,tune,rllib,serve]"
+  ```
+
+### How to setup Ray cluster over spark cluster ?
+
+The architecture of a Spark cluster is as follows:
+![spark-cluster-overview](https://spark.apache.org/docs/latest/img/cluster-overview.png)
+
+Using Spark's barrier mode, we intend to represent each Ray worker node as a long-running Spark
+task, where the Ray worker node has the same set of resources as the Spark task. Multiple Spark
+tasks can run on a single Spark worker node, which means that multiple Ray worker nodes can run
+on a single Spark worker node. Finally, we intend to run the Ray head node on the Spark driver
+node with a fixed resource allocation.
+
+To clarify further, the following example demonstrates how a Ray cluster with 16 total worker CPU
+cores can be launched on a Spark cluster, assuming that each Ray node requires 4 CPU cores and
+10GB of memory:
+
+1. On the Spark driver node, launch the Ray head node as follows:
+
+```
+ray start --head --num-cpus=0
+```
+
+The Ray head node will operate solely as a  manager node, without worker node functionality.
```

Comment:
Note that this is not strictly guaranteed. For example, actors use 0 CPU by default, and they can still be scheduled on this node.